### PR TITLE
Fix/finding #18 Prevent Memory Overwrite in withdrawDepositTo Function

### PR DESCRIPTION
**Pull request by Aboudjem**
This PR addresses the memory overwrite issue identified in finding [#18](https://cantina.xyz/code/d1d4b139-9705-4367-9468-297b7078674e/findings/18). The problem arises in the `withdrawDepositTo` function where the `amount` argument overwrites part of the free memory pointer, leading to potential out-of-gas reverts due to quadratic memory expansion costs. Additionally, this could prevent the actual revert return data from bubbling up.

**Changes Made:**
1. Updated the `withdrawDepositTo` function to store the free memory pointer at the beginning of the assembly block.
2. Modified the return data handling to use the stored free memory pointer, ensuring accurate return data copying and reverting.

### CODE REVIEW

**Comment by livingrockrises** [comment]
the linked issue is wrong ser. #18 goes to github issue number 18

**Comment by livingrockrises** [comment]
shouldn't you/we open an issue on solady repo and validate and discuss this there as well?

**Review by livingrockrises — COMMENTED** [review]
review i

**Review by VGabriel45 — APPROVED** [review]

**Review by VGabriel45 — APPROVED** [review]

File: contracts/base/BaseAccount.sol [hunk 1]
```diff
@@ -74,24 +74,24 @@ contract BaseAccount is IBaseAccount {
...
-            }
-            mstore(0x34, 0) // Restore the part of the free memory pointer that was overwritten.
+/// @notice Withdraws ETH from the EntryPoint to a specified address.
+/// @param to The address to receive the withdrawn funds.
+/// @param amount The amount to withdraw.
+function withdrawDepositTo(address to, uint256 amount) external payable virtual onlyEntryPointOrSelf {
+    address entryPointAddress = _ENTRYPOINT;
+    /// @solidity memory-safe-assembly
+    assembly {
+        let freeMemPtr := mload(0x40) // Store the free memory pointer.
```
Comment (ankurdubey521) [inline contracts/base/BaseAccount.sol:84]: given that this is the first memory allocation of this call frame, I believe it is guaranteed that this value is `0x80`, so we could skip the `mload` and use it directly.

Alternatively, given that the free memory pointer is only being read to store the revert data and that no memory allocations happen post reverting, instead of using `freeMemPtr` with `returndatacopy` and `revert` you could just write everything to `0x0` and assuming `PUSH0` support that would cost 1 gas less.

Comment (Aboudjem) [reply contracts/base/BaseAccount.sol]: fixed

File: contracts/base/BaseAccount.sol [hunk 2]
```diff
@@ -82,14 +82,13 @@ contract BaseAccount is IBaseAccount {
...
+            mstore(0x14, to) // Store the `to` argument directly at memory location 0x14.
+            mstore(0x34, amount) // Store the `amount` argument at memory location 0x34.
+            mstore(0x00, 0x205c2878000000000000000000000000) // Store the function selector at memory location 0x00.
             if iszero(call(gas(), entryPointAddress, 0, 0x10, 0x44, codesize(), 0x00)) {
-                returndatacopy(mload(0x40), 0x00, returndatasize())
-                revert(mload(0x40), returndatasize())
+                returndatacopy(0x0, 0x00, returndatasize()) // Copy return data to 0x0.
+                revert(0x0, returndatasize()) // Revert with data at 0x0.
             }
-            mstore(0x34, 0) // Restore the part of the free memory pointer that was overwritten.
```
Comment (MrToph) [inline contracts/base/BaseAccount.sol:92]: why was this removed? if the call succeeds you still need to reset this from the `mstore(0x34, amount)` above.

File: contracts/base/BaseAccount.sol [hunk 3]
```diff
@@ -82,14 +82,13 @@ contract BaseAccount is IBaseAccount {
         address entryPointAddress = _ENTRYPOINT;
         /// @solidity memory-safe-assembly
         assembly {
-            mstore(0x14, to) // Store the `to` argument.
-            mstore(0x34, amount) // Store the `amount` argument.
-            mstore(0x00, 0x205c2878000000000000000000000000) // `withdrawTo(address,uint256)`.
+            mstore(0x14, to) // Store the `to` argument directly at memory location 0x14.
+            mstore(0x34, amount) // Store the `amount` argument at memory location 0x34.
+            mstore(0x00, 0x205c2878000000000000000000000000) // Store the function selector at memory location 0x00.
```
Comment (ankurdubey521) [inline contracts/base/BaseAccount.sol:87]: Technically the selector is being stored at `0x10`

File: contracts/base/BaseAccount.sol [hunk 4]
```diff
@@ -82,14 +82,13 @@ contract BaseAccount is IBaseAccount {
         address entryPointAddress = _ENTRYPOINT;
         /// @solidity memory-safe-assembly
```
Comment (ankurdubey521) [inline contracts/base/BaseAccount.sol:83]: The following block is no longer memory safe, since it doesn't update the free memory pointer.

Comment (Aboudjem) [reply contracts/base/BaseAccount.sol:83]: good catch

File: contracts/base/BaseAccount.sol (hunk 3 above)
Comment (Aboudjem) [reply contracts/base/BaseAccount.sol]: It was failing so I just reapplied whatever was mentioned in the audit

File: contracts/base/BaseAccount.sol (hunk 2 above)
Comment (livingrockrises) [reply contracts/base/BaseAccount.sol:92]: @MrToph can you please review this again?

**Review by VGabriel45 — APPROVED** [review]

File: contracts/base/BaseAccount.sol (hunk 2 above)
Comment (MrToph) [reply contracts/base/BaseAccount.sol:92]: it is now added back again

**Review by MrToph — APPROVED** [review]
